pennylane_orquestra/orquestra_device.py

```diff
@@ -382,3 +387,254 @@ def qubit_operator_string(self, observable):
+                "{''.join(current_status)}"
+            )
+        return results
+
+    @staticmethod
+    def insert_identity_res_batch(results, empty_obs_list, identity_indices):
+        """An auxiliary function for inserting values which were not computed
+        using workflows into batch results.
+
+        Computations involving the identity observable are given by theoretical
+        values rather than as part of a workflow. Therefore, such values need
+        to be inserted into the results later.
+
+        Args:
+            results (list): workflow results of the batched execution
+            empty_obs_list (list): list of indices where every observable was
+                identity
+            identity_indices (dict): maps the index of a sublist to the
+                the list of indices where the observable was an identity
+
+        Returns:
+            list: list of results
+        """
+        # Insert the lists needed for only identity results
+        for idx in empty_obs_list:
+            results.insert(idx, [])
+
+        # Insert further identity results
+        for list_idx in identity_indices.keys():
+            for iden_idx in identity_indices[list_idx]:
+                results[list_idx].insert(iden_idx, 1)
+
+        return results
+
+    def batch_execute(self, circuits, **kwargs):
+        results = []
+        idx = 0
+        file_prefix = f"{str(uuid.uuid4())}"
+
+        # Iterating through the circuits based on the allowed number of
+        # circuits per workflow
+        while idx < len(circuits):
+            end_idx = idx + self._batch_size
+            batch = circuits[idx:end_idx]
+            file_id = f"{file_prefix}-{str(idx)}"
+
+            res = self._batch_execute(batch, file_id, **kwargs)
+
+            results.extend(res)
+            idx += self._batch_size
+
+        return results
+
+    def _batch_execute(self, circuits, file_id, **kwargs):
+        """Creates a multi-step workflow for executing a batch of circuits.
+
+        Args:
+            circuits (list[QuantumTape]): circuits to execute on the device
+            file_id (str): the file id to be used for naming the workflow file
+
+        Returns:
+            list[array[float]]: list of measured value(s) for the batch
+        """
+        for circuit in circuits:
+            # Input checks
+            not_all_expval = any(obs.return_type is not Expectation for obs in circuit.observables)
+            if not_all_expval:
+                raise NotImplementedError(
+                    f"The {self.short_name} device only supports returning expectation values."
+                )
+
+            self.check_validity(circuit.operations, circuit.observables)
+
+        # 1. Create qasm strings from the circuits
+        # Extract the CircuitGraph object from QuantumTape
+        circuits = [circ.graph for circ in circuits]
+        qasm_circuits = [self.serialize_circuit(circuit) for circuit in circuits]
```

Comment:
Here, I guess the `circuits` is _always_ a tape

Reply:
Yes, as per the original `batch_execute` method (which takes `circuits (list[.tapes.QuantumTape])`). Added a comment to clarify.